### DISCUSSION TITLE
Getting started with VS Code, PlatformIo, ESP32, and ILI9488

### DISCUSSION THREAD

**Discussion by shan-mcarthur** [Bodmer/TFT_eSPI#2555] [General]
I am having difficulties just getting going with this project, and the default patterns seem to be documenting everything I am not using - like Arduino IDE, Unos, and other devices.  What I really want to know is:  what is the easiest way to get a hello world sample going to demonstrate the TFT is working?  Scenario is using Visual Studio Code with PlatformIO extensions and a 3.5 ILI9488 sensor on an ESP32 development board.  I really don't want to go and modify any of the library files and keep my changes local to my project.  I see that there is a platformio.ini file support, and I suspect what I need to do is look at that.

What I am looking for is something along the ideas:
1. How to power the display since the ESP32 is not going to have enough juice and it won't work (so I hear - waiting for my power supply later this week).
2. wire MOSI, SCK, CS, DS, and Reset pins to appropriate pins on your board (and record the pin numbers)
3. wire the LED and VSS to power supply, Ground to ground.
4. create new project (specify a template for hello world, or better yet, the minimal code as it should be small)
5. how to specify the pin settings from step 2 above.

The result should be compile, upload, and presto - it should be showing hello world.

The library is amazingly detailed and very configurable and has super broad support of multiple boards, architectures, communication buses, and TFT screens.  From what I see, this is one of the better libraries ever written.  That said, even though I understand the ease of getting 20 or so samples going by burying the settings into the header files, I think that a better pattern would be appropriate there than asking customers to change it.  There is quite a learning curve until someone can figure it out, and for those of us just building our first LCD project, and don't even know if the hardware is working yet, there are just too many variables.  A simple "get started here to get a Hello World sample going with the least effort" would be very helpful.  Once that is running, geting into more advanced communication protocols, MISO, touch, and Flash could build on that.

**Comment by shan-mcarthur** [comment]
I have it working now.  This wiki page was useful:  https://github.com/Bodmer/TFT_eSPI/wiki/Installing-on-PlatformIO

That said, there are some missing parts that due to the state of the library will show up as a non-functioning screen.  For my use case (an ILI9488 device), I needed to provide additional values.  I believe some of these are necessary for all projects.

Here is the bare minimum project using Visual Studio Code with PlatformIo and using the technique of platformio.ini file modification.  No library file modification needed!

platformio.ini
[env:nodemcu-32s]
platform = espressif32
board = nodemcu-32s
framework = arduino
monitor_speed = 115200
lib_deps = bodmer/TFT_eSPI@^2.5.23
build_flags =
    -D USER_SETUP_LOADED
    -D ILI9488_DRIVER
    -D TFT_MISO=19
    -D TFT_MOSI=23
    -D TFT_SCLK=18
    -D TFT_CS=15
    -D TFT_DC=2
    -D TFT_RST=4
    -D LOAD_GLCD=1
    -D SMOOTH_FONT
    -D SPI_FREQUENCY=27000000

and main.cpp:
#include <Arduino.h>
#include <TFT_eSPI.h> // Hardware-specific library
TFT_eSPI tft = TFT_eSPI();       // Invoke custom library


void setup() 
{
  Serial.begin(115200); // For debug
  tft.init();

  tft.setRotation(1);
  tft.fillScreen(TFT_GREEN);
}

void loop() 
{
  // put your main code here, to run repeatedly:
}

Note that I had to include USER_SETUP_LOADED as well as SMOOTH_FONT.

There are some minor errors in the wiki article.  It projects like we need to have an <macroname>=1 but I found it works just like the include system and is not necessary.

**Reply by foxtesla** [reply]
You have saved my life! ty (L

**Comment by shan-mcarthur** [comment]
I just published an Instructables article on this to help others, and to advocate for them to use this library and participate here in this Github community.  https://www.instructables.com/Using-TFTeSPI-Library-With-Visual-Studio-Code-and-/

**Comment by meilleur102** [comment]
You dont need all this setup. 
The user Setup is there exactly for this. 
The setup 21 will replace all your build flag that you dont need.

**Reply by marcelstoer** [reply]
@meilleur102 is correct here. You could simplify this as described on the wiki page you linked to.

```
  -D USER_SETUP_LOADED=1                        ; Set this settings as valid
  -include $PROJECT_LIBDEPS_DIR/$PIOENV/TFT_eSPI/User_Setups/Setup21_ILI9488.h
```
To my knowledge this works only as long as all your pins are exactly the same as defined in `Setup21_ILI9488.h`. I tried overwriting MISO, MOSI and SCLK like so:

```
  -D USER_SETUP_LOADED=1                        ; Set this settings as valid
  -include $PROJECT_LIBDEPS_DIR/$PIOENV/TFT_eSPI/User_Setups/Setup21_ILI9488.h
  -D TFT_MISO=19
  -D TFT_MOSI=18
  -D TFT_SCLK=05
```
Mixing and matching values from `.h` and PIO build flags didn't work for me.

**Comment by tochinet** [comment]
Yes, this troubled me a lot as well.
IMHO, we should be able to include (and all examples should include as well by default) a line that indicates the setup we wqnt to use, without having to resort to modifying anything in the library files themselves.

As a user of the Arduino IDE platform (that is supposed to be very simple and attractive for people that don't want to edit 100+ char parameters lists for compilers, linkers etc,) I always was convinced that a "sketch" should include a definition of the board it is running on, instead of relying on some immensely long list of boards that you don't have in your lab anyway + speed of Serial, etc.). But this paragraph is a digression ;-)

So, back to TFT_eSPI, I believe that the whole "USER_SETUP" stuff should be reviewed to make it user-friendly, yet it is barely documented. What I'd like to do is simply to add at the beginning of my sketch/program:

```
#include  <TFT_eSPI.h> // Loads the library itself
#define USER_SETUP_ID 25 // Configures the library for the Lilygo T-Display board I want to use
```
Note that the names used should be indicative of the content behind it, so it should better be sth like TFT_eSPI_Lilygo_T-display_Setup.h , but never mind.

**Reply by tochinet** [reply]
Actually, maybe this would not need to be so complex : could the following not do exactly that and allow to be backward compatible ?
```
#define TFT_ESPI_CONFIG25 // Indicates the specific config to select (for the Lilygo T-Display board)
#include  <TFT_eSPI.h> // Loads the library itself
```
And the next part is to replace the include user setup "default" by some case-like selection.
Does it make sense or did I forget sth important ?

**Reply by tochinet** [reply]
Still need to make it work, but actually, it looks like this is already possible, though undocumented. If you want to select a specific board without modifying the library itself (which IMHO is a NO-NO), the program should include the following lines :
```
#include <TFT_eSPI/User_Setups/Setup25_TTGO_T_Display.h>    // Setup file for ESP32 and TTGO T-Display ST7789V SPI bus TFT
#define USER_SETUP_LOADED // Indicates a specific setup is already loaded
#include  <TFT_eSPI.h> // Loads the library itself, it will skip the user select stuff
```
Doesn't work like that, at the moment only the following workaround works : (requires a local copy in your project directory, the full path on the first line won't work either). Any expert can help to correct my mistake ? Arduino-related ?
```
//#include "../libraries/TFT_eSPI/User_Setups/Setup25_TTGO_T_Display.h"  // Setup file for ESP32 and TTGO T-Display ST7789V SPI bus TFT
#include "Setup25_TTGO_T_Display.h"    // Local copy of Setup file for ESP32 and TTGO T-Display ST7789V SPI bus TFT
#define USER_SETUP_LOADED // Indicates a specific setup is already loaded
#include  <TFT_eSPI.h> // Loads the library itself, it will skip the user select stuff
```
